src/aggregator/aggregation/quantile/cm/heap.go

```diff
@@ -34,50 +34,66 @@ func (h minHeap) Min() float64 { return h[0] }
 
 // Push pushes a value onto the heap.
 func (h *minHeap) Push(value float64) {
-	*h = append(*h, value)
-	h.shiftUp(h.Len() - 1)
-}
-
-// Pop pops the minimum value from the heap.
-func (h *minHeap) Pop() float64 {
-	var (
-		old = *h
-		n   = old.Len()
-		val = old[0]
-	)
-
-	old[0], old[n-1] = old[n-1], old[0]
-	h.heapify(0, n-1)
-	*h = (*h)[0 : n-1]
-	return val
-}
-
-func (h minHeap) shiftUp(i int) {
+	if len(*h) == cap(*h) {
+		h.ensureSize()
+	}
+	// append
+	(*h) = append(*h, value)
+	// then, shift up if necessary to fix heap structure. manually inlined.
+	heap := *h
+	n := len(heap)
+	i := n - 1
 	for {
 		parent := (i - 1) / 2
-		if parent == i || h[parent] <= h[i] {
+		if parent == i || heap[parent] <= heap[i] {
 			break
 		}
-		h[parent], h[i] = h[i], h[parent]
+		heap[parent], heap[i] = heap[i], heap[parent]
 		i = parent
 	}
 }
 
-func (h minHeap) heapify(i, n int) {
-	for {
-		left := i*2 + 1
+func (h *minHeap) ensureSize() {
+	heap := *h
+	targetCap := cap(heap) * 2
+
+	newHeap := sharedHeapPool.Get(targetCap)
+	(*newHeap) = append(*newHeap, heap...)
```

Comment:
Doesn't that result in `newHeap` having two sections?
- first (bigger) section with blank values 
- second section with existing values

Reply:
pool resets the slice len to 0 (but it has the right cap):
```

		*v = (*v)[:0]
		return v
	}
makeNew:
	newVal := make(minHeap, 0, size)

	return &newVal
```